crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+            !non_empty_types_scopes.is_empty(),
+            "precondition: there are non-empty types scopes"
+        );
+
+        let (count, scope) = u.choose(&non_empty_types_scopes)?;
+        let count = u32::try_from(*count).unwrap();
+        assert!(!scope.is_empty());
+
+        let max_type_in_scope = u32::try_from(scope.len() - 1).unwrap();
+        let i = u.int_in_range(0..=max_type_in_scope)?;
+
+        Ok(Alias::Outer {
+            count,
+            i,
+            kind: OuterAliasKind::Type,
+        })
+    }
+
+    fn arbitrary_func_type(&mut self, u: &mut Unstructured<'a>) -> Result<FuncType> {
+        let mut params = vec![];
+        let mut param_names = HashSet::new();
+        arbitrary_loop(u, 0, 20, |u| {
+            params.push(self.arbitrary_named_type(u, &mut param_names)?);
+            Ok(true)
+        })?;
+
+        let result = self.arbitrary_value_type(u)?;
+
+        Ok(FuncType { params, result })
+    }
+
+    fn arbitrary_value_type(&mut self, u: &mut Unstructured<'a>) -> Result<ValueType> {
+        let num_choices = if self.current_type_scope().is_empty() {
+            13
+        } else {
+            14
+        };
+        match u.int_in_range(0..=num_choices)? {
+            0 => Ok(ValueType::Unit),
+            1 => Ok(ValueType::Bool),
+            2 => Ok(ValueType::S8),
+            3 => Ok(ValueType::U8),
+            4 => Ok(ValueType::S16),
+            5 => Ok(ValueType::U16),
+            6 => Ok(ValueType::S32),
+            7 => Ok(ValueType::U32),
+            8 => Ok(ValueType::S64),
+            9 => Ok(ValueType::U64),
+            10 => Ok(ValueType::F32),
+            11 => Ok(ValueType::F64),
+            12 => Ok(ValueType::Char),
+            13 => Ok(ValueType::String),
+            14 => {
+                let max_type = u32::try_from(self.current_type_scope().len() - 1).unwrap();
+                Ok(ValueType::Indexed(u.int_in_range(0..=max_type)?))
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn arbitrary_named_type(
+        &mut self,
+        u: &mut Unstructured<'a>,
+        names: &mut HashSet<String>,
+    ) -> Result<NamedType> {
+        let name = crate::unique_string(100, names, u)?;
+        let ty = self.arbitrary_value_type(u)?;
+        Ok(NamedType { name, ty })
+    }
+
+    fn arbitrary_record_type(&mut self, u: &mut Unstructured<'a>) -> Result<RecordType> {
+        let mut fields = vec![];
+        let mut field_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            fields.push(self.arbitrary_named_type(u, &mut field_names)?);
+            Ok(true)
+        })?;
+        Ok(RecordType { fields })
+    }
+
+    fn arbitrary_variant_type(&mut self, u: &mut Unstructured<'a>) -> Result<VariantType> {
+        let mut cases = vec![];
+        let mut case_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            cases.push(self.arbitrary_named_type(u, &mut case_names)?);
```

Comment:
This might be more of a question for the spec, but I thought that the type payload for variants was optional? (only the name would be required)

Reply:
According to the grammars in the binary and explainer docs, the payload is not optional.

Reply:
Hm ok I think that's something to bring up with the spec, I don't think that matches how we've been thinking about variants historically.

Reply:
FWIW, `enum`s are variant types that don't have payloads, and we can always have the unit type payload as well. But, by all means, bring it up with the spec.